<b>CARGA DE LAS LIBRERIAS NECESARIAS

In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error

<b>CARGA DE LOS DATASETS PARA TRABAJAR

In [2]:

reg_1 = pd.read_csv('/datasets/geo_data_0.csv')
reg_2 = pd.read_csv('/datasets/geo_data_1.csv')
reg_3 = pd.read_csv('/datasets/geo_data_2.csv')


<b>REVISAR LOS DATOS QUE CONTIENEN

In [3]:
print(reg_1.head())
print(reg_2.head())
print(reg_3.head())

reg_1.info()
reg_2.info()
reg_3.info()


      id        f0        f1        f2     product
0  txEyH  0.705745 -0.497823  1.221170  105.280062
1  2acmU  1.334711 -0.340164  4.365080   73.037750
2  409Wp  1.022732  0.151990  1.419926   85.265647
3  iJLyR -0.032172  0.139033  2.978566  168.620776
4  Xdl7t  1.988431  0.155413  4.751769  154.036647
      id         f0         f1        f2     product
0  kBEdx -15.001348  -8.276000 -0.005876    3.179103
1  62mP7  14.272088  -3.475083  0.999183   26.953261
2  vyE1P   6.263187  -5.948386  5.001160  134.766305
3  KcrkZ -13.081196 -11.506057  4.999415  137.945408
4  AHL4O  12.702195  -8.147433  5.004363  134.766305
      id        f0        f1        f2     product
0  fwXo0 -1.146987  0.963328 -0.828965   27.758673
1  WJtFt  0.262778  0.269839 -2.530187   56.069697
2  ovLUW  0.194587  0.289035 -5.586433   62.871910
3  q6cA6  2.236060 -0.553760  0.930038  114.572842
4  WPMUX -0.515993  1.716266  5.899011  149.600746
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to

<b>ENTRENAMIENTO Y PRUEBA DEDEL MODELO EN CADA REGION EN GEO_DATA_0.CSV

<b>Separar caracteristicas y objetivo

In [4]:
features = reg_1[['f0', 'f1', 'f2']]
target = reg_1['product']

<B>Dividir los datos en un conjunto de entrenamiento y un conjunto de validacion con una proporcion de 75:25

In [5]:
features_train, features_valid, target_train, target_valid = train_test_split(features, target, test_size=0.25, random_state=12345)

<B>Entrenamiento del modelo

In [6]:
model = LinearRegression()
model.fit(features_train, target_train)

LinearRegression()

<B>Predicciones oara el conjunto de validacion y respuestas correctas del conjunto de valudacion

In [7]:
predictions_valid = model.predict(features_valid)
answers_valid = target_valid

<B>Chequeo rapido

In [8]:
print(predictions_valid[:5])
print(answers_valid.iloc[:5])

[95.89495185 77.57258261 77.89263965 90.17513418 70.51008829]
71751     10.038645
80493    114.551489
2655     132.603635
53233    169.072125
91141    122.325180
Name: product, dtype: float64


<B>Volumen medio de reservas predicho y RMSE del modelo

In [9]:
rmse = mean_squared_error(target_valid, predictions_valid, squared=False)
mean_prediction = predictions_valid.mean()
print('RMSE:', rmse)
print('Volumen medio predicho:', mean_prediction)

RMSE: 37.5794217150813
Volumen medio predicho: 92.59256778438035


<B> DEFINIR FUNCION COMPLETA

In [10]:
def train_and_evaluate(data):
    
    features = data[['f0', 'f1', 'f2']]
    target = data['product']
    
    features_train, features_valid, target_train, target_valid = train_test_split(
        features,
        target,
        test_size=0.25,
        random_state=12345
    )
    
    model = LinearRegression()
    model.fit(features_train, target_train)
    
    predictions_valid = model.predict(features_valid)
    
    rmse = mean_squared_error(target_valid, predictions_valid, squared=False)
    mean_prediction = predictions_valid.mean()
    
    return rmse, mean_prediction

<b>APLICAR FUNCION A CADA UNO DE LOS DATASETS PARA COMPROBARLA Y MOSTRAR LOS RESULTADOS

In [11]:
rmse_0, mean_0 = train_and_evaluate(reg_1)
rmse_1, mean_1 = train_and_evaluate(reg_2)
rmse_2, mean_2 = train_and_evaluate(reg_3)

In [12]:
print('Región 0:')
print('RMSE:', rmse_0)
print('Volumen medio predicho:', mean_0)

print('Región 1:')
print('RMSE:', rmse_1)
print('Volumen medio predicho:', mean_1)

print('Región 2:')
print('RMSE:', rmse_2)
print('Volumen medio predicho:', mean_2)

Región 0:
RMSE: 37.5794217150813
Volumen medio predicho: 92.59256778438035
Región 1:
RMSE: 0.893099286775617
Volumen medio predicho: 68.728546895446
Región 2:
RMSE: 40.02970873393434
Volumen medio predicho: 94.96504596800489


<B>CALCULO DE GANANCIA

<b>Almacenaje de valores necesarios para calculo de variables (definicion de variables)

In [13]:
BUDGET = 100_000_000          
N_WELLS = 200                
REVENUE_PER_UNIT = 4500   

<b>Calculo del umbral minimo por pozo

In [14]:
min_revenue_per_well = BUDGET / N_WELLS
min_units_per_well = min_revenue_per_well / REVENUE_PER_UNIT
print('Ingreso mínimo por pozo:', min_revenue_per_well)
print('Unidades mínimas por pozo:', min_units_per_well)

Ingreso mínimo por pozo: 500000.0
Unidades mínimas por pozo: 111.11111111111111


<b>Guardar resultados de cada region en variables separadas

In [15]:
regions = {
    'Región 0': {'rmse': rmse_0, 'mean_reserves': mean_0},
    'Región 1': {'rmse': rmse_1, 'mean_reserves': mean_1},
    'Región 2': {'rmse': rmse_2, 'mean_reserves': mean_2}
}

<b>Comprar reservas medias con el umbral

In [16]:
for region, values in regions.items():
    print(region)
    print('Reservas medias predichas:', values['mean_reserves'])
    print('Umbral mínimo:', min_units_per_well)
    
    if values['mean_reserves'] >= min_units_per_well:
        print('→ Potencialmente rentable')
    else:
        print('→ Riesgo de pérdidas')

Región 0
Reservas medias predichas: 92.59256778438035
Umbral mínimo: 111.11111111111111
→ Riesgo de pérdidas
Región 1
Reservas medias predichas: 68.728546895446
Umbral mínimo: 111.11111111111111
→ Riesgo de pérdidas
Región 2
Reservas medias predichas: 94.96504596800489
Umbral mínimo: 111.11111111111111
→ Riesgo de pérdidas


<b> Conclusiones

El análisis de las reservas medias predichas para las tres regiones muestra que ninguna alcanza el umbral mínimo de 111.1 unidades por pozo requerido para evitar pérdidas, dado un presupuesto total de 100 millones de dólares para 200 pozos.

Región 0 presenta una reserva media predicha de 92.59 unidades, por debajo del umbral mínimo, lo que indica riesgo de pérdidas si se perforan pozos de manera aleatoria.

Región 1 tiene el valor medio más bajo (68.73 unidades), lo que la convierte en la región menos prometedora desde el punto de vista del promedio de reservas.

Región 2 muestra el valor medio más alto (94.97 unidades), pero aún insuficiente para garantizar rentabilidad promedio.

Estos resultados confirman que la rentabilidad no puede evaluarse únicamente con las reservas medias regionales. Aunque una región pueda parecer más prometedora que otra, la selección de los pozos individuales con mayor volumen esperado es crítica.

Por lo tanto, el siguiente paso consiste en:

seleccionar los 200 pozos con las mayores predicciones de reservas en cada región,

calcular el beneficio esperado basado en estas selecciones,

y evaluar el riesgo de pérdidas mediante simulaciones (bootstrap) antes de tomar una decisión final.

<b>ESCRIBIR UNA FUNCION PARA CALCULAR LA GANACIA DE UN VONJUNTO DE POZOS DE PETROLEO SELECCIONADOS (MODELADO DE PREDICCIONES)

<b>Función para entrenar, predecir y devolver resultados

In [17]:
def get_predictions(data):
    
    features = data[['f0', 'f1', 'f2']]
    target = data['product']
    
    features_train, features_valid, target_train, target_valid = train_test_split(
        features,
        target,
        test_size=0.25,
        random_state=12345
    )
    
    model = LinearRegression()
    model.fit(features_train, target_train)
    
    predictions = model.predict(features_valid)
    
    return pd.Series(predictions), target_valid.reset_index(drop=True)

<b>Funcion para calcular ganacia

In [18]:
def calculate_profit(predictions, target):
    top_predictions = predictions.sort_values(ascending=False).head(N_WELLS)
    real_reserves = target[top_predictions.index]
    total_reserves = real_reserves.sum()
    profit = total_reserves * REVENUE_PER_UNIT - BUDGET
    return profit, total_reserves

<b>Ejecutar para cada region

In [19]:
regions = {
    'Región 0': reg_1,
    'Región 1': reg_2,
    'Región 2': reg_3
}
results = {}

for region, data in regions.items():
    predictions, target_valid = get_predictions(data)
    profit, reserves = calculate_profit(predictions, target_valid)
    
    results[region] = {
        'profit': profit,
        'reserves': reserves
    }

<b>Mostrar resultados

In [20]:
for region, values in results.items():
    print(region)
    print('Reservas totales:', values['reserves'])
    print('Ganancia potencial ($):', values['profit'])

Región 0
Reservas totales: 29601.83565142189
Ganancia potencial ($): 33208260.43139851
Región 1
Reservas totales: 27589.081548181137
Ganancia potencial ($): 24150866.966815114
Región 2
Reservas totales: 28245.22214133296
Ganancia potencial ($): 27103499.635998324


<b>CALCULO DE RIESGOS Y GANANCIAS PARA CADA REGION

<b>Constantes

In [21]:
BOOTSTRAP_SAMPLES = 1000

<b>Funcion para calcular ganacias

In [22]:
def calculate_profit(predictions, target):
    top_predictions = predictions.sort_values(ascending=False).head(N_WELLS)
    real_reserves = target[top_predictions.index]
    profit = real_reserves.sum() * REVENUE_PER_UNIT - BUDGET
    return profit

<b>Funcion del bootstrapping

In [23]:
def bootstrap_profit(predictions, target):
    profits = []
    
    for _ in range(BOOTSTRAP_SAMPLES):
        subsample = predictions.sample(n=len(predictions), replace=True, random_state=None)
        subsample_target = target[subsample.index]
        
        profit = calculate_profit(subsample, subsample_target)
        profits.append(profit)
    
    return pd.Series(profits)

<b>Ejecutarlo para cada region

In [24]:
bootstrap_results = {}

for region, data in regions.items():
    predictions, target_valid = get_predictions(data)
    profits = bootstrap_profit(predictions, target_valid)
    
    bootstrap_results[region] = profits

<b>Calculo de metricas finales

In [25]:
for region, profits in bootstrap_results.items():
    mean_profit = profits.mean()
    ci_lower = profits.quantile(0.025)
    ci_upper = profits.quantile(0.975)
    loss_risk = (profits < 0).mean() * 100
    
    print(region)
    print('Ganancia promedio:', mean_profit)
    print('IC 95%:', (ci_lower, ci_upper))
    print('Riesgo de pérdidas (%):', loss_risk)

Región 0
Ganancia promedio: 166426948.84455144
IC 95%: (135801348.51142097, 203043500.8710424)
Riesgo de pérdidas (%): 0.0
Región 1
Ganancia promedio: 148173858.54065442
IC 95%: (120988543.2009309, 183063976.68433845)
Riesgo de pérdidas (%): 0.0
Región 2
Ganancia promedio: 152929084.1830465
IC 95%: (124690168.9890432, 189549441.9601804)
Riesgo de pérdidas (%): 0.0


<b>Conclusiones

Los resultados muestran que las tres regiones presentan un riesgo de pérdidas del 0%, lo que indica una alta estabilidad del modelo al seleccionar los 200 pozos con mayor volumen predicho.

Sin embargo, al comparar las ganancias promedio y los intervalos de confianza, se observan diferencias claras:

Región 0 presenta la mayor ganancia promedio (≈ 166.4 millones de dólares) y el intervalo de confianza más alto.

Región 2 ocupa el segundo lugar en ganancia promedio (≈ 152.9 millones de dólares).

Región 1 muestra la ganancia promedio más baja (≈ 148.2 millones de dólares).

Dado que la Región 0 combina la mayor ganancia esperada con un riesgo de pérdidas nulo, se propone esta región como la mejor candidata para el desarrollo de pozos petrolíferos.